### DISCUSSION TITLE
Headless output

### DISCUSSION THREAD

**Discussion by robbins** [niri-wm/niri#714] [Ideas]
Does Niri have the ability to create headless outputs that aren't tied to any physical monitor? e.g. for programs like Moonlight/Sunshine?

e.g. https://wiki.archlinux.org/title/Sway#Create_headless_outputs

**Comment by YaLTeR** [comment]
Not at the moment.

**Comment by djmaze** [comment]
Not only for that, it is also needed for virtual desktops in VR / MR (which I am planning to do in the future). People reported successfully using headless outputs in Hyprland for use in the _Immersed_ virtual desktop environment.

I have no deeper knowledge about this topic, but this sounds to me like the used smithay toolkit already somewhat supports these kind of usecases: https://github.com/Smithay/smithay/issues/287

**Reply by YaLTeR** [reply]
I don't think anything is missing on the Smithay side for headless outputs.

**Reply by djmaze** [reply]
Talking about _immersed_, I just tried running it (on NixOS). On sway it works, but on niri it fails to start, outputting this message:

```
listener function for opcode 2 of wl_output is NULL
```

What could be the reason? Is this maybe a client problem [as similarly reported here](https://github.com/swaywm/sway/issues/6713)?

**Reply by djmaze** [reply]
Ah yes, client problem, just found this: https://github.com/pop-os/cosmic-epoch/issues/1137

**Comment by FlyingWombat** [comment]
If you make support for headless output, please consider including a way to detach an existing session from its display and make it headless. A use case for this is starting a VM with VFIO gpu passthrough, without having to kill the user session.

**Reply by YaLTeR** [reply]
Interesting use case, could you link some more info about it?

**Comment by Quackdoc** [comment]
I have found a usecase for headless, I have found myself wanting to test niri on all sorts of different interaction methods and one of the more convenient ways of doing this is using streaming solutions such as sunshine + moonlight. 

The goal would be to create a headless session at whatever output is desired, then stream that to whatever interface I am testing on. This isn't the nicest solution in the world, but it should work fine.

**Reply by Quackdoc** [reply]
As an appendage to this, not sure if this is a "headless" part or not, but I did find myself in a usecase where it would be nice to have a "pipewire only" output mode. I think this would be a "headless + always on pipewire". The case here is remote streaming use case where I only view the output over a network via gstreamer.

**Comment by isti115** [comment]
One more really nice use case that I've been relying on with other wayland compositors (Sway and River) is to repurpose other devices as extra monitors by piping the contents of headless outputs to theirs screens through VNC.

**Comment by hcoohb** [comment]
Oh, I would love this feature as well!

My use case is basically we have a media server at home, which output is constantly driving the TV and i am also remotely accessing it as a different user. I currently use a virtual x11 display coupled with VNC to use for everything as my laptop is not performant.

But it would be so much better to use niri with a "virtual" output!

**Comment by Hikiru** [comment]
i dont have a second monitor, and sunshine/moonlight with a headless display is a great way to use one of my other devices as one.

**Comment by myNicole98** [comment]
I second this, it would be really nice to have a headless display option in Niri

**Comment by envicious** [comment]
Third this! I love using vnc as a nice virtual monitor whenever i need some extra space :)
Would love to do this in Niri too

**Comment by zitongcharliedeng** [comment]
+1 for NixOS e2e tests

**Comment by clarkarch** [comment]
Any update on this? i need this for sunshine and moonlight streaming

**Comment by Vibricks** [comment]
Would love to also have this for using my devices as a second monitor with sunshine + moonlight

**Comment by QaidVoid** [comment]
I have implemented it on my fork: https://github.com/QaidVoid/niri/tree/feat/virtual. It's not polished but it works for my use case. Tested with moonlight/sunshine and wayvnc.

> Disclaimer: most of this was written with the help of AI

**Reply by repparw** [reply]
Will you make a PR?

**Reply by QaidVoid** [reply]
I'm not 100% confident if the implementation is correct. It might be better if someone more knowledgeable picks it up.

**Reply by repparw** [reply]
someone will review it, better than start from scratch surely

**Reply by patroza** [reply]
sadly crashes on a laptop when closing the lid of my laptop.

**Comment by juliadin** [comment]
I'd love to see headless niri for running wayvnc without local graphics output

**Comment by subsy** [comment]
Another vote here - sunshine/moonlight being my primary use case

**Comment by aaronlauterer** [comment]
Definitely another vote here.
One more use case I have for headless displays is when giving presentations online where you have to share the screen. I don't want to or cannot share my main screen, as it is a high resolution widescreen monitor.
So the resolution would not work because of the aspect ratio and resolution. Additionally, switching a PDF or presentation tool into full screen on the main screen is a problem when you also need it for the video call.

**Reply by YaLTeR** [reply]
Windowed fullscreen + dynamic cast target set to the window. https://niri-wm.github.io/niri/Screencasting.html#windowed-fakedetached-fullscreen

**Reply by aaronlauterer** [reply]
> Windowed fullscreen + dynamic cast target set to the window. https://niri-wm.github.io/niri/Screencasting.html#windowed-fakedetached-fullscreen

Thanks! I wasn't aware of that approach.

Unless I am missing something else, for more complex scenarios, like live demos or screen recordings involving multiple windows/applications, the headless display would be a great addition. As one could then set the resolution to a better-fitting one, usually full HD. :-)

**Comment by roufpup** [comment]
Would love to see this feature added for use with sunshine

**Comment by vinnavannewton** [comment]
it's very useful for me, so im casting a vote!

**Comment by rektide** [comment]
Sizable write-up for virtual outputs in #3101.

**Comment by OG-LORD-RK** [comment]
I'd really like to see this become a native Niri feature.

My main use case is using **Sunshine + Moonlight as a second monitor**. What I'm looking for isn't just streaming/mirroring an existing output — I'd like Niri to expose a virtual output that behaves like another monitor, so I can move windows to it normally.

Something as simple as an IPC action to create/remove a virtual output would be really useful, e.g.:

niri msg action create-virtual-output

with the ability to specify things like resolution, refresh rate, scale, and output name.

I noticed the virtual-output fork mentioned above and that it has already been tested with Sunshine/Moonlight. That's pretty much exactly the functionality I'm looking for.

It would be great if that work could eventually make its way into upstream Niri so we don't need a separate compositor fork just to use a virtual monitor.

For me the ideal setup would basically be:

Niri > virtual output > Sunshine > Moonlight

with the remote device behaving as a proper extended display rather than a stream of the existing monitor.

**Comment by subsy** [comment]
FWIW, you can have this today without a compositor fork, but the trick is a layer below niri, in the kernel's DRM stack, so niri never knows the monitor isn't real.

Take an unused connector on your GPU (I use an empty `DP-1` on an AMD card), inject a synthetic EDID, and force it "connected":

```bash
dir=/sys/kernel/debug/dri/*/DP-1
cat /usr/lib/firmware/edid/my-1080p.bin > $dir/edid_override
echo on > $dir/force
echo 1 > $dir/trigger_hotplug    # amdgpu-specific re-probe
```

The kernel then reports the connector as connected, niri picks it up as a genuine output and applies whatever output "DP-1" { ... } block you've configured. It is a real monitor to niri: real workspaces, move-column-to-monitor-* works, windows move onto it normally. Then point a second Sunshine instance at it - note that under capture = wlr, output_name matches the connector name (DP-1), not a numeric index, and you'll want a separate port and file_state so it doesn't collide with your main instance.

I've been running exactly this for months with the extra display configured to my non-standard FW 13 laptops resolution so it appears native from there. And I just made another one at 1080p for a separate sunshine instance that I use for streaming games to our living room Apple TV.

That said, the reasons I'd still very much like this native in niri:

- Root and debugfs. It's a system-level unit poking /sys/kernel/debug, not something the user's session owns. It doesn't belong in dotfiles and doesn't follow the session around.
- Driver-specific. trigger_hotplug is an amdgpu thing. On other drivers you're into video=DP-1:e on the kernel command line or drm_kms_helper.edid_firmware, i.e. a reboot.
- You must have a spare physical connector, and you burn a real CRTC on it. No good on a laptop with one output, or in a VM.
- The mode list is frozen by the EDID blob. If Moonlight asks for a different resolution or refresh rate, you can't honour it - you regenerate the blob and re-hotplug. This is the big one (for me at least): a native virtual output could be created at the client's requested resolution!!
- The monitor exists permanently, scanning out at 60 Hz whether or not anyone is streaming, and windows wander onto a screen nobody is looking at - sometimes can be quite annoying.

That last pair is what an IPC action would really unlock. With niri msg action create-virtual-output / destroy-virtual-output you could put them straight into Sunshine's apps.json as a prep-cmd do/undo pair - the display is created at exactly the resolution and refresh the client negotiated when the stream starts, and vanishes when it ends. No phantom monitor, no wasted vblanks, no root, no EDID hex editing, and it works on any GPU or in a VM.

## What native support would mean for us specifically

1. **On-demand displays at the client's resolution.** Today the Apple TV gets 1080p60 because that's what's baked into the EDID blob. With a create/destroy IPC action wired into Sunshine's `prep-cmd`, the output would be born at whatever the client actually negotiated and die when the stream ends. That's the thing the hack fundamentally cannot do.
2. **No permanent phantom monitor.** Right now on my system there's a screen at x=4288 that exists 24/7. Windows land on it, the overview shows it, and the GPU scans it out at 60 Hz for nothing.
3. **It becomes dotfile-managed.** The script and the system unit are root-owned and outside chezmoi, so a fresh machine doesn't get them from a `chezmoi apply` — it's a manual step you have to remember. An IPC action is just a line in `autostart.kdl` or `apps.json`.
4. **Portability.** The current approach is tied to an AMD card with a spare DP connector. Native support would work on the laptop, in a VM, or on a box with every connector already in use.

The suggested alternative (windowed fullscreen plus a dynamic cast target) solves the *presentation* case but not this one. There's no output, so there's nowhere to move windows to.

**Comment by juliadin** [comment]
And in Addition to this usecase (a virtual additional monitor for whatever application) I want to reiterate that I would love to run niri fully headless as well ... no physical display, just a virtual output with wayvnc. Just throwing that back in here because a lot of the discussion revolves around additional virtual displays

**Comment by OG-LORD-RK** [comment]
Thanks! I’ve actually tried that approach before, but it’s not particularly
practical for my setup. I’m using a laptop because I travel quite a lot for
university, and I usually use an Android tablet as my second screen when I
need one.

So carrying additional hardware just for this isn’t really ideal for me,
especially since I don’t currently have the extra hardware needed for that
setup anyway. I’d rather wait for proper virtual display support in niri
than build a workaround around it.

Appreciate the suggestion though!
